File: native-modules/expo-sherpa-kws/android/src/main/java/expo/modules/sherpakws/ExpoSherpaKwsModule.kt
package expo.modules.sherpakws

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ExpoSherpaKwsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoSherpaKws")

    Events("onKeywordDetected")

    AsyncFunction("startKWS") { modelDir: String, keywordsFile: String, promise: Promise ->
      // TODO: Initialize sherpa-onnx keyword spotter via JNI
      // - Load models from modelDir
      // - Start AudioRecord feeding to spotter
      println("[ExpoSherpaKws] startKWS: $modelDir")
      promise.resolve(null)
    }

    AsyncFunction("stopKWS") { promise: Promise ->
      // TODO: Stop audio recording and destroy spotter
      println("[ExpoSherpaKws] stopKWS")
      promise.resolve(null)
    }

    AsyncFunction("enrollSpeaker") { audioSamples: List<Double>, promise: Promise ->
      // TODO: Extract speaker embedding via JNI, store securely
      println("[ExpoSherpaKws] enrollSpeaker: ${audioSamples.size} samples")
      promise.resolve(true)
    }

    AsyncFunction("verifySpeaker") { audioSamples: List<Double>, promise: Promise ->
      // TODO: Extract embedding, cosine compare
      println("[ExpoSherpaKws] verifySpeaker")
      promise.resolve(mapOf("passed" to true, "score" to 1.0))
    }

    AsyncFunction("getEnrollmentStatus") { promise: Promise ->
      // TODO: Check secure storage for enrollment
      promise.resolve(false)
    }
  }
}

File: native-modules/expo-sherpa-kws/ios/ExpoSherpaKwsModule.swift
import ExpoModulesCore

public class ExpoSherpaKwsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoSherpaKws")

    Events("onKeywordDetected")

    AsyncFunction("startKWS") { (modelDir: String, keywordsFile: String) in
      // TODO: Initialize SherpaOnnxKeywordSpotter with C API
      // - Load encoder/decoder/joiner from modelDir
      // - Load keywords from keywordsFile
      // - Start AVAudioEngine tap feeding audio to spotter
      print("[ExpoSherpaKws] startKWS called with modelDir: \(modelDir)")
    }

    AsyncFunction("stopKWS") {
      // TODO: Stop audio engine tap and destroy spotter
      print("[ExpoSherpaKws] stopKWS called")
    }

    AsyncFunction("enrollSpeaker") { (audioSamples: [Double]) -> Bool in
      // TODO: Use SherpaOnnxSpeakerEmbeddingExtractor
      // - Extract embedding from audioSamples
      // - Average with existing enrollments
      // - Store in Keychain/SecureStore
      print("[ExpoSherpaKws] enrollSpeaker called with \(audioSamples.count) samples")
      return true
    }

    AsyncFunction("verifySpeaker") { (audioSamples: [Double]) -> [String: Any] in
      // TODO: Extract embedding, cosine compare with enrolled
      // - Return { passed: Bool, score: Float }
      print("[ExpoSherpaKws] verifySpeaker called")
      return ["passed": true, "score": 1.0]
    }

    AsyncFunction("getEnrollmentStatus") { () -> Bool in
      // TODO: Check if enrollment embedding exists in storage
      return false
    }
  }
}
